src/ripple/basics/SlabAllocator.h

```diff
@@ -57,40 +57,66 @@ class SlabAllocator
         // A linked list of appropriately sized free buffers:
         std::uint8_t* l_ = nullptr;
 
-        // The next memory block
+        // The next block:
         SlabBlock* next_;
 
         // The underlying memory block:
-        std::uint8_t const* const p_ = nullptr;
+        std::atomic<std::uint8_t*> data_ = nullptr;
 
-        // The extent of the underlying memory block:
-        std::size_t const size_;
+        // The number of items allocated from this block
+        std::atomic<std::size_t> avail_{0};
 
-        SlabBlock(
-            SlabBlock* next,
+        SlabBlock(SlabBlock* next) : next_(next)
+        {
+        }
+
+        ~SlabBlock()
+        {
+            // We don't release the allocated memory here: slabs are only
+            // destroyed when the process is shutting down, and we cannot
+            // guarantee that this isn't called before the destruction of
+            // all the objects that are allocated in the buffer we have.
+        }
+
+        std::uint8_t*
+        assign(
             std::uint8_t* data,
             std::size_t size,
-            std::size_t item)
-            : next_(next), p_(data), size_(size)
+            std::size_t item,
+            std::size_t count)
         {
-            // We don't need to grab the mutex here, since we're the only
-            // ones with access at this moment.
+            assert(
```

Comment:
I don't really object, but our code style is to put function implementations outside of class definitions. I understand this file already did it this, so I don't mind if you keep it, but thought I'd point it out.

Reply:
I get it, and I could definitely move the functions below the classes if you feel strongly enough about it, but I think the better option here (given that this is somewhat esoteric, unlikely to be casually edited, etc) is to simply inline the function implementations directly into the class.

Reply:
I'm fine with it as-is. Just thought I'd point it out.